### PR TITLE
Use AES GCM for kernel encryption to ensure integrity

### LINKED ISSUE #1: crypto: encryption is not authenticated
The current crypto implemented aes256-cfb with a magic is not secure against [bit flipping attacks](https://en.wikipedia.org/wiki/Bit-flipping_attack) which is a very real vector for boot image protections like these.

Modes like CFB and anything in `aes::cipher` are what is called "confidentiality only" modes. There are two things you want to protect

- Confidentiality: any part of the plaintext should not get out
- Integrity: any outside tampering would not go undetected

In case of CFB you could just flip the i-th bit in ciphertext and the decryption will happily go though with the i-th bit in plaintext flipped. If you know what you are doing and are good at making educated guess on where to alter the boot image would lead to takeover.. The boot image will decrypt to whatever you want it to look like

To fix this there are two main solutions: [HMAC](https://en.wikipedia.org/wiki/HMAC) which is essentially a hash-secured message seal and [AEAD ciphers](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Authenticated_encryption_with_additional_data_(AEAD)_modes) which adds secure check bits throughout the ciphertext to make sure it is has not been tampered with.

Usually with HMAC you get a constant size increase (256 bits for sha256) but with AEAD you have a small but linear increase in size.
HMAC is more "leak proof" but there are [not really](https://eprint.iacr.org/2015/477.pdf) effective ways to tamper AEAD protected ciphertext [given you used them correctly ofc](https://www.elttam.com/blog/key-recovery-attacks-on-gcm/) either.

It looks like you are just trying to protect a relatively small blob that you will decrypt fully in memory before use. I think for simplicity we should go with an HMAC based protection, now replacing the "magic" with the HMAC digest during encryption. During decryption the user puts in the password we derive the key and then use the key to redigest the encrypted boot image. Compare the two digests if they match then continue decrypting.

Or if you want to go full on with AEAD modes we could use gcm and xts, looks like RustCrypto has [implementations](https://github.com/RustCrypto/AEADs/tree/master) for them as well in case in future you want to decrypt blocks larger than memory or O(1) random access

In the future if we want to go full on configurable you could make something like this and then parse parameters from string, for example `aes256-cfb-magic` (not authenticated) `aes256-gcm-none` (authenticated) `aes256-cfb-sha256` (authenticated), like the whole deal in veracrypt, LUKS, etc. This would ofc need more planning and might be a future goal out of the scope of this issue.

```rust
pub struct Algorithm {
    cipher: Cipher,
    cipher_mode: CipherMode,
    hmac: Option<Hmac>,
}
```

Let me know your thoughts and I could get a PR started!

### PR DESCRIPTION
**Pull request by MarcusGrass**
Fixes https://github.com/MarcusGrass/boot-rs/issues/1.
Encrypts kernel using AES GCM to ensure integrity of the decrypted payload.

Not a lot of meaningful changes outside of the IV/nonce going from 16 to 12 bytes, some drive-by clippy cleanups.

### CODE REVIEW

File: boot-lib/src/crypt.rs [hunk 1]
```diff
@@ -85,48 +84,47 @@ pub fn derive_with_salt(
...
+pub fn encrypt(
+    src: &[u8],
+    key: &[u8; REQUIRED_HASH_LENGTH],
+    iv: [u8; REQUIRED_IV_LENGTH],
+) -> Result<Vec<u8>, String> {
     let key: &GenericArray<u8, U32> = GenericArray::from_slice(key);
-    //let cipher = aes::Aes128::new(key);
-    Aes256Cfb8Enc::new(key, &iv.into()).encrypt(&mut src);
-    src
+    let nonce = Nonce::from_slice(&iv);
```
Comment (eternal-flame-AD) [inline boot-lib/src/crypt.rs]: Nonce should not come from a static config file, this is a footgun for the user bc if nonce is unchanged between encryptions [key recovery is possible](https://www.elttam.com/blog/key-recovery-attacks-on-gcm/). I used [GCM-SIV](https://docs.rs/aes-gcm-siv/latest/aes_gcm_siv/) mode is more resilient to these kinds of mistakes at a little cost of performance although the expectation from users is the same: nonce are what they mean they should be random

Take a look at my implementation, which hides the complexity of choosing IVs and nonces from the user.

[Here](https://github.com/eternal-flame-AD/boot-rs/blob/c6d0fdfbce6210dab9297ee5d84ec9f214a538e5/boot-crypt/src/algorithm.rs#L146), he encryption key and necessary IV/nonce is simply derived from one Argon2 KDF, I wrote a `fixed_iv` override just for backward compatibility. During encryption the salt is randomly [generated](https://github.com/eternal-flame-AD/boot-rs/blob/c6d0fdfbce6210dab9297ee5d84ec9f214a538e5/boot-crypt/src/parcel.rs#L38) and written into the blob [header](https://github.com/eternal-flame-AD/boot-rs/blob/c6d0fdfbce6210dab9297ee5d84ec9f214a538e5/boot-crypt/src/parcel.rs#L17), guaranteeing that IV and nonce never repeat. During decryption we just parse the header and use the same salt to recover the cipher state.

With this the user only have to do [this](https://github.com/eternal-flame-AD/boot-rs/blob/c6d0fdfbce6210dab9297ee5d84ec9f214a538e5/boot-crypt/src/lib.rs#L22) to encrypt and seal data, and the input is only plaintext and password.

File: boot.cfg.tst
Comment (eternal-flame-AD) [inline boot.cfg.tst:1]: remove the IV and salt config from this file for reason above

**Review by eternal-flame-AD — CHANGES_REQUESTED** [review]
Looks good except the nonce issue. I would definitely distinguish IV and nonce, it is renamed to nonce because it MUST be random for each encryption stream.

I also tried myself to build a misuse resilient and configurable crypto system [here](https://github.com/eternal-flame-AD/boot-rs/tree/c6d0fdfbce6210dab9297ee5d84ec9f214a538e5/boot-crypt), where all the user has to do is give a password , I generate a random salt internally and the instantiation with the IVs and nonces just happen. The salt is written to the file header so during decryption I can internally recover the IV/nonce and the user can just use the data using the `unseal` and then `decrypt` methods.

**Comment by MarcusGrass** [comment]
Sorry for the delay, it's bin a lot lately, now the encryption details are baked into the encrypted blob header!

File: Cargo.lock [hunk 2]
```diff
@@ -1,7 +1,17 @@
-# This file is automatically @generated by Cargo.
+git# This file is automatically @generated by Cargo.
```
Comment (eternal-flame-AD) [inline Cargo.lock]: What's happening here? 'D

File: boot-strap/src/main.rs [hunk 3]
```diff
@@ -230,28 +210,31 @@ fn generate(opts: &GenBootOpts) -> Result<(), String> {
...
     // Insanity check.
+    let encrypted = encrypted?;
     if encrypted == kernel_data {
         return Err(
             "Encryption failed, output data same as input data (Sanity check triggered)."
                 .to_string(),
         );
     }
-    let mut enc_c = encrypted.clone();
+    let enc_c = encrypted.clone();
```
Comment (eternal-flame-AD) [inline boot-strap/src/main.rs]: You don't need to clone now

File: boot-lib/src/crypt.rs [hunk 4]
```diff
@@ -1,240 +1,273 @@
...
     #[test]
-    fn encrypt_decrypt() {
-        let key = [0u8; 32];
-        let iv = [0u8; 16];
-        let data = b"My spooky data";
-        let mut encrypted = encrypt(data.to_vec(), &key, iv);
-        assert_ne!(data.as_slice(), &encrypted);
-        let decrypted = decrypt(&mut encrypted, &key, iv);
-        assert_eq!(data.as_slice(), decrypted);
+    fn serde_test() {
```
Comment (eternal-flame-AD) [inline boot-lib/src/crypt.rs]: metadata_serialize_deserialize or metadata_serdes

File: boot-strap/src/main.rs [hunk 5]
```diff
@@ -301,12 +279,13 @@ fn generate(opts: &GenBootOpts) -> Result<(), String> {
 }
 
 #[inline]
-fn generate_crypt_randoms() -> Result<([u8; 16], [u8; 32]), String> {
```
Comment (eternal-flame-AD) [inline boot-strap/src/main.rs:304]: maybe rename this function?? It does much more than it says on the name

**Review by eternal-flame-AD — COMMENTED** [review]
Looks good, dunno why you decided to keep the nonce out of the KDF, I would say a bit more complex than it has to be but okay. Just that one fat finger and good to go!

In the long run I suggest 
1. maybe refactor some bin programs, there are super long functions called thing like generate, gen_key_file which I find it a bit hard to read, at least use empty lines to give it some separation on what subtask you are doing.
1. would be cool to have some documentations on how to set this thing up from a regular LUKS encrypted install. Call me wrong if I what I am saying is impossible, I would think it would be easier if you don't take away the job of currently well-known facilities like mkinitcpio and just write hooks for that instead.

File: Cargo.lock (hunk 2 above)
Comment (MarcusGrass) [reply Cargo.lock]: I think this was a slip that somehow made it into the commit, it was overwritten locally so I didn't even notice!

File: boot-strap/src/main.rs (hunk 3 above)
Comment (MarcusGrass) [reply boot-strap/src/main.rs]: Nice catch, clippy usually swoops these up!

**Comment by MarcusGrass** [comment]
I'll settle for this for now, having both the salt and nonce in the header is unnecessary, but I think I'll have to rethink the header anyway. Ideally it would have a flexible length so that the header could be parsed in a backwards compatible way if stuff changes in it. I was thinking about duplicating the header as a footer to prevent bit-rot making the kernel unbootable, then again maybe that's not the biggest issue when it's just a single file that can fairly easily be rebuilt.  

Ideally I'll figure out a way not to include any configuration into to the bootloader binary, but for that I have to find a way to navigate the vfat filesystem on the UEFI-level, if I can do that, the build step would be a lot simpler, initially, an initramfs.cfg could be generated, then any time you change the kernel you just run `boot-strap` without needing to recompile the bootloader, that way it could more easily go into a mkinitcpio or whatever else.

Funnily, the project structure is messy because it started as just an initramfs in Rust, then I started looking at the bootloader, then merged the projects, so it's a bit strange overall.

It probably should be separate things, but it's difficult to separate if encryption is present, at least without looking at decrypting a luks partition. That might not necessarily be so hard, but I haven't looked into it yet. It has to be done at the UEFI-level, which is always a bit of a hassle. 

I think the main improvements that would make this usable is to figure out if I can find `cwd` in UEFI, if I can do that we don't need any data inserted into the bootloader binary. Secondly, figuring out how to pass kernel parameters when starting the UEFI-image is important to be able to support an initramfs that's not compiled into the kernel. 

But as always, first make it work, then make it good. I think having correct encryption is prio 1, thank you a lot for the help with that!

**Pull request merged**